### DISCUSSION TITLE
Provide a slot for custom extension in the collection/layout view

### DISCUSSION THREAD

**Discussion by Kematia** [directus/directus#3841] [Feature Requests]
Hey! 

I was thinking about offering a way to plug in a custom extension on the collection layout, which could be used to manipulate the items provided to the layout (table/cards).

## My use case

![image](https://user-images.githubusercontent.com/17851386/106270267-fd70a380-622d-11eb-9ff8-7af0c822cc7c.png)
![image](https://user-images.githubusercontent.com/17851386/106273370-d2d51980-6232-11eb-9800-f673fc276d72.png)

_Don't judge my code, this was a weekender POC and it's not in production_ 😂
 ___
I have a collection of "products", with a 3 levels deep filter via a dynamic dropdown, which searches for a normalized path in a related collection, or filters products with a certain tag, or shows products missing some custom values.

As of now it works correctly, but to create this custom filtering method I had to manually plug in the "table" layout, and it's a bit tricky to maintain and extend. 

## 80/20
While it might not fit exactly in the 80/20 rule, I think being able to apply custom logic to the items passed to the layout will greatly extend the "pluggability" of the collection layout, which I reckon is a central part of the application. 

Advanced filters are very powerful, but lacks some UX to really be handled by clients/editors.

For example, in my use case the status is displayed as a localized label (Bozza), but in the db the value is saved in english (draft) - or I have a custom display to parse a JSON object, which is not handled (IIRC) by the advanced filters.

Seeing that I'm fine with the default layouts, or in case I'm also ok with detaching the filtering logic from a custom layout*, I would love to have a place to pass my own extension to apply my custom logic/UX, before the items hit the layout.

I'm open to feedback & willing to work on this issue**!

___
_\* I use the same component in the item view to set the value.
\** it looks to me this is deep in the bowels of Directus, and I might need some pointers!_

**Comment by Kematia** [comment]
Addendum: this concept could be expanded on a per field basis, here's a use case:

![image](https://user-images.githubusercontent.com/17851386/106317066-0719fb80-626e-11eb-9082-6afa16c5af08.png)


I have a repeater to input a price field (label/value pair), which can either be a single row: 
![image](https://user-images.githubusercontent.com/17851386/106317099-100acd00-626e-11eb-87fa-1db81b38e9cd.png)

or a more complex array of price/labels, which are the same for every product (wine bottles in this case)

![image](https://user-images.githubusercontent.com/17851386/106317170-33357c80-626e-11eb-814c-f94f23757b8d.png)

I had to wrap (and setup) a repeater field with two buttons, that populate it with the array of prices.

**Comment by rijkvanzanten** [comment]
After a quick meetup with @Kematia, we had a follow up idea:

Theoretically speaking, you can achieve rendering custom buttons etc around existing interfaces by making your own custom interface, and then render the existing interface in _that_. However, for this kind of functionality (having a button that pre-sets the value to one of multiple defaults), it would be super helpful to be able to render your own items in the field menu:

![CleanShot 2021-01-29 at 15 45 19](https://user-images.githubusercontent.com/9141017/106325668-3407e780-6249-11eb-9437-685d6afc099c.png)

Thoughts on that @benhaynes? You'd be able to render your own menu items in there for things like "Pre-fill template" or "Check with Grammarly" or "Translate on Crowdin"

**Reply by benhaynes** [reply]
Yeah! I'm down to include these new "extension types". It seems like we could have a dynamic extension type that could inject into any `slots` in the App. That way we can create slots for the header bar, field label dropdown, sidebar, etc... it's like event hooks, but for the App!

**Reply by danielbidala** [reply]
@benhaynes This would be an incredibly powerful feature. For example add custom buttons (copy) to collection details, right before delete button without creating a new module or layout.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
